Add state sync support

erikgrinaker opened the pull request:
Fixes #5689, fixes #5690. Adds support for taking, restoring, and serving snapshots for [state sync](https://github.com/tendermint/tendermint/issues/828), as outlined in [ADR-053](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-053-state-sync-prototype.md) and the [ABCI reference](https://docs.tendermint.com/master/spec/abci/apps.html#state-sync).

* `rootmulti.Store` now implements a new `snapshots.Snapshotter` interface, which can take and restore binary state snapshots at a given height. Snapshots are built as follows:

  1. `rootmulti.Store` iterates over all `iavl.Store` stores, and for each one emits a Protobuf `SnapshotStoreItem` message with store metadata

  2. For each `iavl.Store`, export `iavl.ExportNode` nodes via `Export()` at the given height and emit a Protobuf `SnapshotIAVLItem` message for each node

  3. The length-prefix framed Protobuf message stream is zlib-compressed

  4. The zlib-compressed stream is chunked naïvely into 10 MB chunks

* `snapshots.Store` stores snapshot metadata in a separate database and chunks in a filesystem directory

* `BaseApp` takes asynchronous snapshots after `Commit()` in regular height intervals given by the config option `state-sync.snapshot-interval` (default 0, i.e. disabled). It also prunes old snapshots, and keeps the most recent snapshots given by the option `state-sync.snapshot-keep-recent` (default 2). The snapshot interval must be a multiple of `pruning-keep-every`, to make sure IAVL versions aren't deleted while the height is being snapshotted.

* `BaseApp` implements the state sync ABCI interface, for fetching and applying state snapshots.

* `SimApp` has been updated to enable and initialize state sync snapshotting.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [x] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Review `Codecov Report` in the comment section below once CI passes

fedekunze commented on store/types/types.proto:
```diff
@@ -0,0 +1,28 @@
+syntax = "proto3";
+package cosmos_sdk.store.v1;
```
nit, this should live in the `proto/` directory

erikgrinaker replied:
Good call, thanks - this has changed a bit since the PR was first submitted back in March. :)

fedekunze commented on snapshots/types/types.proto:
```diff
@@ -0,0 +1,20 @@
+syntax = "proto3";
```
ditto move to `proto/`

marbar3778 commented on server/config/config.go:
```diff
@@ -91,6 +91,17 @@ type GRPCConfig struct {
+// StateSyncConfig defines the state sync snapshot configuration.
+type StateSyncConfig struct {
+	// SnapshotInterval sets the interval at which state sync snapshots are taken.
+	// 0 disables snapshots. Must be a multiple of PruningKeepEvery.
```
Will this change in tandem with the pruning settings? If I stop my node change the pruning strategy then restart without changing this field what will happen?

erikgrinaker replied:
You'll get an error. The snapshot interval must be a multiple of the pruning keepever interval.

This is a somewhat arbitrary limitation. Snapshotting just needs a guarantee that heights won't be removed while being snapshotted, and this is currently the only way the SDK can give this guarantee. If the snapshotter could request the pruner to delay pruning of a height until the snapshot is complete then these settings no longer need to match.

alexanderbez replied:
Let's create a followup issue for that @erikgrinaker. This isn't a blocker and what you have now should suffice in most cases.

erikgrinaker replied:
Added #7188 for this.

marbar3778 commented on server/config/config.go:
```diff
@@ -156,6 +168,10 @@ func DefaultConfig() *Config {
 			Address: DefaultGRPCAddress,
 		},
+		StateSync: StateSyncConfig{
+			SnapshotInterval:   0,
```
I think this should be a multiple of the default pruning strategy. It is best to default snapshots intervals because this will be something that is overlooked by most node operators. This could lead to very few nodes in a network able to provide snapshots.

erikgrinaker replied:
Yeah, I don't have a strong opinion here, so I'll let the SDK team decide on a default.

alexanderbez replied:
We could pick a very relaxed option here so it happens very infrequently so we can get some testing. Like every 1k, 5k or 10k blocks

erikgrinaker replied:
Yeah, but if we do that then _all_ nodes in the network are on the same schedule. If there's a bug, _all_ nodes fail, and the entire network goes down.

alexanderbez replied:
Fair point, let's keep at 0 then?

robert-zaremba replied:
How does it impact the storage and performance?

Is 0 === disabled?

erikgrinaker replied:
0 is disabled, yes. Taking a snapshot will cause a lot of IO, and the entire state at the snapshotted version (i.e. not including historical versions) will be stored in compressed form. If I remember correctly, when I tried this with the Cosmos Hub state back in February, it was less than 100 MB of data and took less than a minute to snapshot.

robert-zaremba commented on snapshots/manager.go:
```diff
@@ -0,0 +1,223 @@
+// beginLocked begins an operation while already holding the mutex.
+func (m *Manager) beginLocked(op operation) error {
+	if op == opNone {
+		return errors.New("can't begin a none operation")
```
ditto, this is also an internal logic error, would be good to mark it with a specific type.

robert-zaremba commented on snapshots/manager.go:
```diff
@@ -0,0 +1,223 @@
+		}
+		m.endLocked()
+		return err
+	case <-time.After(20 * time.Millisecond):
```
why `time.After` is added?

erikgrinaker replied:
We need to know if the goroutine above (`m.target.Restore()`) had an error, so we have to wait for it to be scheduled. This is a real code smell, so if you have a better suggestion I'm all ears. One option would be for `Restore()` to return a `Restorer` struct that we could feed chunks to, and error if the snapshot is invalid, but that seemed a bit much.

robert-zaremba replied:
Other idea is to create another channel or a semaphore and pass it to the `Restore` function. Once it will be ready for listening, we will send a signal (channel) or open the semaphore. I think channel is easier here because you have a switch.

erikgrinaker replied:
Yeah, I added a ready channel:

```go
Restore(height uint64, format uint32, chunks <-chan io.ReadCloser, ready chan<- struct{}) error
```

It's not super-clean, but it works and is fairly lightweight. Would possibly be better to return a restorer, in the same way as IAVL's `MutableTree.Import()` returns an importer, but it makes the interface harder to implement and I'd like to keep it lightweight.

robert-zaremba commented on snapshots/manager.go:
```diff
@@ -0,0 +1,223 @@
+	defer m.mtx.Unlock()
+	if m.operation != opRestore {
+		return false, fmt.Errorf("no restore operation in progress")
+	}
```
how do we know that we are still in the same _restore_ process?

Here is the use-case:

1. Restore started, let's call this process `A` (with process I'm referring to a whole flow, network process, not a goroutine)
1. We received few chunks.
1. in `store.Restore` something fails and we close the `chDone`
1. we start second Restore process (`B`) - this will initialize a new object to `m.chRestoreDone`
1. some node from process `A` will keep sending chunks, then

    1. the `RestoreChunk` function will be called with data from process `A`, but having everything (including `m.chRestoreDone` and `m.chRestore` initialized for process `B`
    2. `store.Restore` will receive data from process `A`
    3. it will probably fail because data doesn't match and we are in a kind of a dead loop.

erikgrinaker replied:
We don't, it's up to the caller to keep track of this. We could return a `Restorer` for the restore operation that the caller has to pass chunks to, but I'm not sure if it's necessary.

erikgrinaker replied:
Also, in your example above, if process A fails, why would it keep sending chunks? It should just exit on error. In this case, the process is Tendermint, so it would be a bug in Tendermint.

robert-zaremba replied:
Process A is in the new node who is performing a restore. 
A node, which offered a snapshot in process A can still send the snapshots, even though A failed. And that chunks will land in B, correct?

erikgrinaker replied:
Tendermint controls the restore process. It will take care of fetching chunks, ordering them, queueing them, and feeding them to the application. If something goes wrong, Tendermint will attempt to start a new restore process. The SDK currently does not support this, since I it's not able to reset its IAVL stores to a fresh state before starting a new restore. This is something we can implement later.

robert-zaremba replied:
I've created this issue: https://github.com/cosmos/cosmos-sdk/issues/7192
Let's add a comment here to mark this error as an `internal` error.

robert-zaremba commented:
I have few architecture questions (sorry, I'm still learning the details about the system):

1. Are we always able to correctly determine the number of chunks? Would it be better to have it dynamic (eg, have an async message saying chunk X is the last one)?
2. Where is the lightclient check in the code (is it in this PR or it's in separate issue?)
    * How we know that lightclient is up-to-date with the data from the latest snapshot? Does lightclient implement the full tendermint consensus algorithm?

robert-zaremba commented:
Shouldn't we somehow let the user decide from which snapshot he wants to sync? For example Algorands' Fast Catchup requires a snapshot ID / hash  when starting a node with a fast catchap mode.

https://developer.algorand.org/docs/run-a-node/setup/install/#sync-node-network-using-fast-catchup

erikgrinaker commented on baseapp/abci.go:
```diff
@@ -358,6 +385,100 @@ func (app *BaseApp) Query(req abci.RequestQuery) abci.ResponseQuery {
+		// We currently don't support resetting the IAVL stores and retrying a different snapshot,
+		// so we ask Tendermint to abort all snapshot restoration.
+		return abci.ResponseOfferSnapshot{Result: abci.ResponseOfferSnapshot_ABORT}
+	}
```
ABCI responses generally do not return error messages: https://docs.tendermint.com/master/spec/abci/abci.html#errors

erikgrinaker commented on baseapp/baseapp_test.go:
```diff
@@ -105,6 +110,64 @@ func setupBaseApp(t *testing.T, options ...func(*BaseApp)) *BaseApp {
+
+	app := setupBaseApp(t, append(options,
+		SetSnapshotStore(snapshotStore),
+		SetSnapshotInterval(2),
```
This is a test fixture. If we use a larger number (e.g. 10) then the tests will take much longer, since we need to wait for additional blocks to be produced before we get the necessary number of snapshots needed by the test.

erikgrinaker commented on simapp/simd/cmd/root.go:
```diff
@@ -188,6 +200,9 @@ func newApp(logger log.Logger, db dbm.DB, traceStore io.Writer, appOpts serverty
 		baseapp.SetIndexEvents(cast.ToStringSlice(appOpts.Get(server.FlagIndexEvents))),
+		baseapp.SetSnapshotStore(snapshotStore),
+		baseapp.SetSnapshotInterval(cast.ToUint64(appOpts.Get(server.FlagStateSyncSnapshotInterval))),
+		baseapp.SetSnapshotKeepRecent(cast.ToUint32(appOpts.Get(server.FlagStateSyncSnapshotKeepRecent))),
```
I believe this is a typical pattern used e.g. by gRPC to pass options to the server. I personally prefer option structs, but just following what's already there.

alexanderbez replied:
ditto (commented in the issue). I'd prefer to keep this as-is.

erikgrinaker commented on snapshots/alias.go:
```diff
@@ -0,0 +1,20 @@
+	Snapshotter = types.Snapshotter
+	Snapshot    = types.Snapshot
+	Metadata    = types.Metadata
+)
```
This was following a pattern used by the SDK when the PR was initially submitted back in March. It doesn't seem to be used anymore, so I'll drop this and just use the `types` sub-package directly.

alexanderbez replied:
yeah we're dropping alias -- I was never a fan of it.

erikgrinaker replied:
Removed the aliasing, so callers will instead have to import `snapshots/types`. Not a huge fan of this convention, but it's what the SDK seems to use elsewhere - let me know if you'd like anything to be done differently.

erikgrinaker commented on snapshots/helpers_test.go:
```diff
@@ -0,0 +1,144 @@
+)
+
+func checksum(b []byte) []byte {
+	hash := sha256.Sum256(b)
```
These are mostly used as checksums, and cannot be trusted anyway (the snapshot metadata is off-chain and thus can be forged by anyone). Light client verification is used for the final verification, this is just a best-effort checksum to guard against IO issues and non-determinism.

robert-zaremba replied:
If you don't care about preimage attack resistance then don't use cryptographic hash. It's an overkill. Murmur3 hash or Googles' fast-hash are much better for some checksum. However they are easy to attack (if you want to find different data with same checksum).

erikgrinaker replied:
We're already using SHA-256 for everything else (including consensus-critical stuff), so I figured we might as well just use that here too. But if you'd like to use a different algo just let me know and I'll swap it.

robert-zaremba replied:
If we don't need preimage resistance then yes - let's use `murmur3`. Using `sha256` for hashing many big chunks of data is overkill.

erikgrinaker replied:
Sure. We don't need preimage resistance for safety (the light client verification handles that), but I suppose it could protect against DoS attacks - if an adversary can provide a known-bad chunk that matches the hash of a valid snapshot then they could crash the recovery process, preventing the node from state syncing.

robert-zaremba replied:
That should be a slashing behavior. But I guess it won't work because not every full node is bonded / active validator.

I think even with sha2 a node can send bad data and it will be only noticed after de-serialization, correct? So cryptography hash function doesn't help.

erikgrinaker replied:
Yes, collisions are possible with every hash function, it's just a matter of probability/difficulty.

robert-zaremba replied:
I didn't talk about collisions. I said that even with `sha2` a node can send a bad data:

1. Node A (bad) is creating a chunk, do some random change and then hashes it with `sha2`.
2. Node B (victim) is reconstructing a chunk, hashes it. Hash check is correct.
3. Node B tries to restore the chunk - it will eventually crash because it's broken.

erikgrinaker replied:
Ah, but we don't necessarily get the chunk hashes from A, they come from the snapshot manifest.

When starting state sync, Tendermint will discover snapshots in the network, by fetching manifests that include chunk hashes (in the SDK's case). It then determines which snapshot is "best" and downloads chunks from all nodes that claim to have this snapshot. If a node can generate a faulty chunk with the same hash listed in the manifest they can cause the node to fail.

The heuristic for "best" snapshot is currently very naïve (pick the newest one), but should be improved to use the one available on a majority of nodes (weighted by age). In this scenario, an adversary would have to control a majority of snapshotting nodes to be able to send a faulty chunk, but this no longer holds if the hash is vulnerable to preimage attacks.

robert-zaremba replied:
OK, thanks for explanation. 
So, let's keep it as is and create a new issue to see what will be the best hash function for this.

robert-zaremba replied:
https://github.com/cosmos/cosmos-sdk/issues/7215

erikgrinaker commented on snapshots/helpers_test.go:
```diff
@@ -0,0 +1,144 @@
+		hasher.Write(chunk)
+	}
+	return hasher.Sum(nil)
+}
```
This is just a test helper, so I don't think performance is particularly critical here.

robert-zaremba replied:
I know it's a test, bit I think it's good in general to have a habit to reuse resources if possible in bigger systems

erikgrinaker replied:
I think tests should optimize for developer convenience, not resource usage - especially something as trivial as this.

erikgrinaker commented on snapshots/store.go:
```diff
@@ -0,0 +1,370 @@
+	snapshotHasher := sha256.New()
+	for chunkBody := range chunks {
+		defer chunkBody.Close() // nolint: staticcheck
+		chunkHasher := sha256.New()
```
The cost of creating the hasher is most likely negligible compared to the cost of doing the snapshot IO, so I don't think it matters, but sure.

robert-zaremba replied:
It's just moving the line and calling a reset

erikgrinaker replied:
Done

erikgrinaker commented on snapshots/store.go:
```diff
@@ -0,0 +1,370 @@
+
+	bHeight := make([]byte, 8)
+	binary.BigEndian.PutUint64(bHeight, height)
+	k = append(k, bHeight...)
```
Nice trick

erikgrinaker commented on snapshots/store_test.go:
```diff
@@ -0,0 +1,366 @@
+		}
+	}
+	return store, teardown
+}
```
I'm not sure table-driven tests would be beneficial for these sorts of tests, since they depend on a relatively large amount of state and the test logic can be complex. We'd basically end up with a big switch/case for each of the test cases, in which case there's no real point to it.

erikgrinaker replied:
Also, the code you're referring to is just a fixture, not a test. It sets up the pre-existing state for all the other tests.

erikgrinaker commented on snapshots/manager.go:
```diff
@@ -0,0 +1,223 @@
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	if m.operation != opRestore {
+		return false, fmt.Errorf("no restore operation in progress")
```
Sure, I'll be happy to follow whatever convention the SDK uses. Do you have an example of this elsewhere?

robert-zaremba replied:
I don't think there is one. I've joined Regen recently. I'm happy to to propose such structure. So probably it's better to do it in a separate PR , and here we can mark it with some special TODO. Let me know - if OK I will create a new issue.

erikgrinaker replied:
Sure, fine with me - I'm not really too involved with SDK development, just did this since I was building the Tendermint-side of state sync anyway.

robert-zaremba replied:
There is already a structure for this errors in SDK. 

Please use `errors.New(codespace string, code uint32, desc string)` (a package from cosmos-sdk) for all errors.
You can add there a new variable for the codespace: `TendermintCodespace` if it's relevant for some errors here.   And we need to add a code for `LogicErrors`.

erikgrinaker replied:
> Please use `errors.New(codespace string, code uint32, desc string)` (a package from cosmos-sdk)

You sure about that? `sdkerrors.New` is not used anywhere else in the code base, the convention seems to be to register base error types in `types/errors/errors.go` via `sdkerrors.Register`, and then use those errors together with `sdkerrors.Wrap` to add messages. Would you like me to add a type `ErrLogic` and use this with `sdkerrors.Wrap`?

robert-zaremba replied:
You are right. Good idea with `ErrLogic` + `Wrap`. Let's do so. In other places let's also use `Wrap` whenever possible to add some types to the errors.

erikgrinaker replied:
Done, added `ErrLogic` and `ErrConflict`

erikgrinaker commented on snapshots/manager.go:
```diff
@@ -0,0 +1,223 @@
+	case err := <-m.chRestoreDone:
+		if err == nil {
+			err = errors.New("restore ended unexpectedly")
+		}
```
Sure, that makes sense.

erikgrinaker replied:
Done, although I personally feel like this made the code less clear.

erikgrinaker commented on snapshots/manager.go:
```diff
@@ -0,0 +1,223 @@
+
+	// Pass the chunk to the restore, and wait for completion if it was the final one.
+	m.chRestore <- ioutil.NopCloser(bytes.NewReader(chunk))
+	m.restorePending = m.restorePending[1:]
```
Yeah, either that, or just pre-create a buffered channel containing all of the hashes.

erikgrinaker commented on snapshots/util.go:
```diff
@@ -0,0 +1,162 @@
+// NewChunkReader creates a new ChunkReader.
+func NewChunkReader(ch <-chan io.ReadCloser) *ChunkReader {
+	return &ChunkReader{ch: ch}
+}
```
Ideally we wouldn't need to use byte slices at all, and only use IO streams, but unfortunately ABCI does not support streaming. I built it this way so that we could implement a streaming API at some later point (this is quickly becoming a necessity for e.g. `InitChain`, where we currently on chain upgrades have to load the entire state into memory, which isn't going to work for much longer).

Another advantage is that `io.ReadCloser` allows us to propagate IO errors, via `pipe.CloseWithError()`. If we pass byte slices we would have to wrap them in another result struct for error propagation as well.

robert-zaremba replied:
OK, maybe we can add a comment line explaining that.

erikgrinaker replied:
Yup, done

erikgrinaker commented:
> Are we always able to correctly determine the number of chunks? Would it be better to have it dynamic (eg, have an async message saying chunk X is the last one)?

Since snapshots are pre-generated on a fixed schedule, we always know ahead of time how many chunks a snapshot has. This is an optimization to speed up restores, since actually taking the snapshot is the time-consuming part, so fetching and restoring a pre-generated snapshot is much faster.

In the initial protocol, each chunk had a boolean indicating that it was the final chunk, but users complained that this was unnecessary and confusing (since we already give the number of chunks in the snapshot metadata), so it was dropped.

>Where is the lightclient check in the code (is it in this PR or it's in separate issue?)

Tendermint does the light client verification after the snapshot has been restored, by checking the app hash against the on-chain app hash. You can read some more details on this from the Tendermint side here:

https://docs.tendermint.com/master/spec/abci/apps.html#state-sync

Note that we don't do any incremental verification of chunks against the app hash, as this was considered out of scope for an initial version. Since Cosmos Hub restores take about 30 seconds anyway, erroring early doesn't really save that much time.

> How we know that lightclient is up-to-date with the data from the latest snapshot? Does lightclient implement the full tendermint consensus algorithm?

The light client is given a root of trust (a height and commit hash), and then finds the trusted on-chain app hash for the snapshot height based on that via untrusted RPC servers. Details on the light client protocol here:

https://medium.com/tendermint/everything-you-need-to-know-about-the-tendermint-light-client-f80d03856f98

>Shouldn't we somehow let the user decide from which snapshot he wants to sync? For example Algorands' Fast Catchup requires a snapshot ID / hash when starting a node with a fast catchap mode.

In the current protocol, Tendermint does snapshot discovery and picks the snapshot that seems "best". We could have the user find and specify a specific snapshot instead, but that would need e.g. block explorer interfaces to expose those snapshots to the user, and make it harder to use. Open to discussing this though, feel free to open a Tendermint issue with a proposal.

erikgrinaker commented on snapshots/types/snapshotter.go:
```diff
@@ -0,0 +1,14 @@
+
+	// Restore restores a state snapshot.
+	Restore(height uint64, format uint32, chunks <-chan io.ReadCloser) error
+}
```
Yeah, personally I'd like to get rid of the `/types` subdir as well, I just find them really unergonomic - but just following the code base convention.

erikgrinaker commented on baseapp/abci.go:
```diff
@@ -324,6 +330,27 @@ func (app *BaseApp) halt() {
 	os.Exit(0)
 }
 
+// snapshot takes a snapshot of the current state and prunes any old snapshots.
```
Sure, changed to "create" everywhere.

erikgrinaker commented on baseapp/abci.go:
```diff
@@ -358,6 +385,100 @@ func (app *BaseApp) Query(req abci.RequestQuery) abci.ResponseQuery {
+
+	snapshots, err := app.snapshotManager.List()
+	if err != nil {
+		app.logger.Error("Failed to list snapshots", "err", err.Error())
```
Good call, updated

robert-zaremba reviewed (changes requested):
Thanks for all updates. good job.

I've checked in Cosmos SDK about error handling - there is a type for that and we just need to add a code for logic errors and maybe tendermint codespace. Please look at the comment I left.

erikgrinaker commented:
Thanks for reviewing! I think I've addressed all of your comments apart from the error handling (see [separate response](https://github.com/cosmos/cosmos-sdk/pull/7166#discussion_r480272632)), is there anything else outstanding?

robert-zaremba commented:
> Thanks for reviewing! I think I've addressed all of your comments apart from the error handling (see [separate response](https://github.com/cosmos/cosmos-sdk/pull/7166#discussion_r480272632)), is there anything else outstanding?

@erikgrinaker  Looks very good -- let's close the parts about errors (I've just responded to it) and hashing.

erikgrinaker commented:
I believe @alexanderbez is reviewing this as well, so I'll wait for his approval.

alexanderbez reviewed (approved):
utACK. Overall, looks solid! I left a few minor remarks, but otherwise it LGTM 👍

zmanian commented:
This seems so close to merging that I'm excited to have it `stargate-2`

erikgrinaker commented on baseapp/baseapp.go:
```diff
@@ -60,6 +63,11 @@ type BaseApp struct { // nolint: maligned
 
+	// manages snapshots, i.e. dumps of app state at certain intervals
+	snapshotManager    *snapshots.Manager
+	snapshotInterval   uint64 // block interval between state sync snapshots
+	snapshotKeepRecent uint32 // recent state sync snapshots to keep
```
In that case the manager would have to do the scheduling. Now, the baseapp figures out when to take snapshots (i.e. `if height%snapshotInterval==0`) and then calls `snapshotManger.Create()`. If the manager should do this, we would have to add a method `snapshotManager.MaybeCreate()` that would be called for every single committed height.

Considered implementing it that way originally, but thought it might be a bit odd - happy to change it.

erikgrinaker replied:
Ah, remembered why I didn't: then the snapshotManager would have to have a logger (to emit e.g. "Creating snapshot" messages). When we keep the policy outside, the manager doesn't have to do any logging, which is usually indicative of a cleaner design.

alexanderbez replied:
I don't totally follow why having the manager have its own logger is a design flaw or poor design choice. But given this doesn't actually buy us anything, it's fine to leave as-is then.

erikgrinaker replied:
It couples the manager more tightly with the rest of the system. But I agree it's not really a big deal in this instance, either approach works for me.

erikgrinaker commented on proto/cosmos/base/store/v1beta1/snapshot.proto:
```diff
@@ -0,0 +1,28 @@
+syntax = "proto3";
+package cosmos.base.store.v1beta1;
+
+import "gogoproto/gogo.proto";
```
Each `Snapshotter` implementation is free to define its own binary snapshot encoding - the `snapshots` package does not care about what snapshots contain, only how to handle them. So the encoding follows the implementer, in this case `rootstore`.

erikgrinaker commented on snapshots/manager.go:
```diff
@@ -0,0 +1,259 @@
+			"a more recent snapshot already exists at height %v", latest.Height)
+	}
+
+	chunks, err := m.target.Snapshot(height, types.CurrentFormat)
```
When we change the snapshot format, we could conceivably want nodes to be able to generate snapshots both in the old and new format, to support both old and new state sync clients. Currently this is entirely academic, since we only have a single format - can hardcode it if you'd prefer.

alexanderbez replied:
This is fine for now then.

erikgrinaker commented on types/errors/errors.go:
```diff
@@ -123,6 +123,14 @@ var (
+
+	// ErrConflict defines a conflict error, e.g. when two goroutines try to access
+	// the same resource and one of them fails.
+	ErrConflict = Register(RootCodespace, 36, "conflict")
```
These are the current scenarios that trigger `ErrConflict`:

* The BaseApp tries to start a snapshot/prune operation while a different snapshot/prune operation is in progress (e.g. when creating a snapshot takes longer than the snapshot interval).

* A snapshot is created for a height that already has a snapshot.

* A snapshot is attempted created or deleted while it is being created.

Conflict is the usual nomenclature for these sorts of concurrency issues. We could call it e.g. `ErrRace`, but I'm not sure if that's any clearer, and it's not always a race condition (e.g. block replay can cause the application to attempt to create a snapshot that has already been created, which isn't a race, simply non-idempotency).

alexanderbez replied:
Totally get that, but when I see "conflict", I don't automatically get that context. I don't think this is expressive enough to infer that. Unless you already know this, "conflict" could mean anything.

erikgrinaker replied:
Ah - since this is always wrapped with an error message, the context sort of follows from that. But maybe I'm using this error convention wrong. Would you prefer having individual error types for these cases? I'm not really sure what the alternative would be.

alexanderbez replied:
No, you're right. I took a look at all the uses of `ErrConflict` and the context is sufficient. I would say we could rename this to `SnapshotConflict`, but we may end up having to use a similar error for other parts of the codebase in the future, so `ErrConflict` is OK 👍 

resolving...

robert-zaremba replied:
I prefer `ErrConflict`. The context about Snapshot can be added in a wrapper.

alexanderbez commented:
Lets fix the merge conflicts and get this bad boy merged.

erikgrinaker commented:
Awesome, thanks for reviewing, I know it was a fairly big chunk! Will resolve and merge first thing tomorrow!
